src/ipc/helper-ipc3.c

```diff
@@ -875,8 +876,9 @@ int ipc_comp_free(struct ipc *ipc, uint32_t comp_id)
 			icd->cd->pipeline->sched_comp = NULL;
 	}
 
-	/* free component and remove from list */
-	comp_free(icd->cd);
+	/* free component and remove from list, used the allocated cached address */
+	if (icd->cd->is_shared)
+		comp_free((struct comp_dev *)uncache_to_cache(icd->cd));
```

Comment:
Lets keep it simple. The allocator APIs should accept cache/uncached and internally convert to the correct alias. The Zephyr API by default returns uncached addresses and it's up to the user to convert to cacheble.

Reply:
> Lets keep it simple. The allocator APIs should accept cache/uncached and internally convert to the correct alias. The Zephyr API by default returns uncached addresses and it's up to the user to convert to cacheble.

Hi @lgirdwood , the context of the problem is like this:
a. the buffers are always created on core 0.
b. the components are created on its specific core with cached(0xbexxxxxx) address.
c. when connecting buffer to a secondary core component with ipc_buffer_to_comp_connect() or ipc_comp_to_buffer_connect(), the component address will be mapped to a uncached(0x9exxxxxx) one with comp_make_shared().
d. all the subsequent usage of the component will use the uncached address.
e. at comp_free(), freeing the component memory with uncached address will throw error as this 0x9exxxxxx address is not managed by the current Non-Zephyr allocator.

Reply:
@lgirdwood NAK on having the allocator convert between cached and uncached. The allocator itself should be as platform-independent as possible (the heaps themselves should be the only thing that is platform dependent), and having it do this conversion on its own is against this design decision.

Holding the original address for this purpose sounds good to me, just like how we're already doing with allocating aligned memory.

Reply:
@paulstelian97 what I mean here is that I should be able to 
```c
free(0xbe100000); // cached alias
free(0x8e100000); // uncached alias
```
and the allocator should convert internally so it free the correct block object. This means application code does not need to care about the pointer alias.

Furthermore, cachable memory should be an "opt-in" for clients (as clients have extra work), we should always return uncachle alias if supported by default except for buffers (like Zephyr does).  

I expect Zephyr to manage all kernel objects wrt cache on your new SMP HW, but I think SOF objects will be per core (i.e. pipeline1 is only on core 1 - so no cache management is needed as no pipeline1 code should run on core 0). This really simplifies our core code (as we can direct IPCs in the handler to the correct core) and minimize cache awareness to only a small number of core logic files.

Reply:
@lgirdwood Is having such aliases cached vs uncached memory required if we want to do SMP for future hardware revisions? The current one doesn't have one and cache_to_uncache/uncache_to_cache macros are implemented as no-ops for the code to build correctly. I didn't keep track if the number of uses of these macros got lowered or if they got away from core code to be fair...

Reply:
@paulstelian97 I think the aliasing is cheap for HW on xtensa. See the cache_attr values in the linker script. Your HW guys would need to create the alias and you would need to set the cache_attr correctly in the linker script. I would recommend this approach for simplicity and performance.

Reply:
@keyonjie are you able to do the following ?

> what I mean here is that I should be able to
> 
> ```c
> free(0xbe100000); // cached alias
> free(0x8e100000); // uncached alias
> ```
> 
> and the allocator should convert internally so it can free the correct block object. This means application code does not need to care about the pointer alias.

Reply:
> @paulstelian97 what I mean here is that I should be able to
> 
> ```c
> free(0xbe100000); // cached alias
> free(0x8e100000); // uncached alias
> ```
> 
> and the allocator should convert internally so it can free the correct block object. This means application code does not need to care about the pointer alias.
> 
> Furthermore, cachable memory should be an "opt-in" for clients (as clients have extra work), we should always return uncachle alias if supported by default except for buffers (like Zephyr does).
> 
> I expect Zephyr to manage all kernel objects wrt cache on your new SMP HW, but I think SOF objects will be per core (i.e. pipeline1 is only on core 1 - so no cache management is needed as no pipeline1 code should run on core 0). This really simplifies our core code (as we can direct IPCs in the handler to the correct core) and minimize cache awareness to only a small number of core logic files.

That's almost exact, but we do have some tricky parts, some IPC will be handled only on Core 0, e.g. dai_config hw configuration part, some memories need to be shared among Cores, e.g. DMA channels, DMIC configurations, ...

For the issue here, the 'is_shared' flag was added to try to handle component level multi-core support, so it is another exception of the assumption that "pipeline1 is only on core 1", so we need to either remove the 'is_shared' logic, or adding the fix like what I do here to do the opposite transforming, or, to make the allocator intelligent enough as @lgirdwood mentioned.